File: tests/Unit/MatchServiceTests.swift
import Foundation
import XCTest

@MainActor
final class MatchServiceTests: XCTestCase {
    func testPersistAndRestoreActiveMatch() async throws {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        XCTAssertNotNil(store.active)
        XCTAssertEqual(store.active?.currentGame.leftPoints, 1)

        let restored = MatchService(store: store)
        XCTAssertEqual(restored.activeMatch?.currentGame.leftPoints, 1)
        XCTAssertEqual(restored.activeMatch?.id, service.activeMatch?.id)
    }

    func testFileStoreRoundTrip() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = FileMatchStore(directory: dir)
        let engine = ScoringEngine()
        var match = engine.startMatch()
        match = try engine.apply(.selectServer(.left), to: match)
        match = try engine.apply(.pointWon(.right), to: match)
        try store.saveActiveMatch(match)

        let loaded = try store.loadActiveMatch()
        XCTAssertEqual(loaded?.currentGame.rightPoints, 1)

        match = try engine.apply(.endEarly, to: match)
        try store.saveActiveMatch(nil)
        try store.archiveMatch(match)
        let archive = try store.loadArchivedMatches()
        XCTAssertEqual(archive.count, 1)
        XCTAssertEqual(archive.first?.status, .endedEarly)
    }

    func testDiscardDoesNotArchive() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        service.discardMatch()
        XCTAssertNil(service.activeMatch)
        XCTAssertTrue(service.archivedMatches.isEmpty)
        XCTAssertNil(store.active)
    }

    func testDiscardBeforeServerSelectionClearsActiveMatch() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        XCTAssertEqual(service.activeMatch?.isWaitingForFirstServe, true)
        service.discardMatch()
        XCTAssertNil(service.activeMatch)
        XCTAssertTrue(service.archivedMatches.isEmpty)
        XCTAssertNil(store.active)
    }

    func testEndEarlyArchivesAndKeepsActiveUntilAck() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        service.endMatchEarly()
        XCTAssertEqual(service.activeMatch?.status, .endedEarly)
        XCTAssertEqual(service.archivedMatches.count, 1)
        service.acknowledgeCompletedMatch()
        XCTAssertNil(service.activeMatch)
        XCTAssertEqual(service.archivedMatches.count, 1)
    }

    func testUndoViaService() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        service.awardPoint(to: .right)
        service.undoLastPoint()
        XCTAssertEqual(service.activeMatch?.currentGame.leftPoints, 1)
        XCTAssertEqual(service.activeMatch?.currentGame.rightPoints, 0)
    }

    func testSelectingServerUpdatesActiveMatch() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        XCTAssertEqual(service.activeMatch?.needsServerSelection, true)
        service.selectServer(.right)
        XCTAssertEqual(service.activeMatch?.needsServerSelection, false)
        XCTAssertEqual(service.activeMatch?.currentServer, .right)
    }

    func testCompleteWarmUpPersistsAndKeepsServePrompt() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        XCTAssertEqual(service.activeMatch?.needsWarmUp, true)
        XCTAssertEqual(service.activeMatch?.needsServerSelection, true)
        service.completeWarmUp()
        XCTAssertEqual(service.activeMatch?.needsWarmUp, false)
        XCTAssertEqual(service.activeMatch?.needsServerSelection, true)
        XCTAssertEqual(store.active?.needsWarmUp, false)
    }

    func testNewServeAtTheChangeoverPromptsAndPersists() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        var settings = MatchSettings.default
        settings.setsToWin = 2
        service.startMatch(settings: settings)
        service.selectServer(.left)
        for _ in 0..<24 {
            service.awardPoint(to: .left) // 6-0, set to Us
        }
        XCTAssertEqual(service.activeMatch?.completedSets.count, 1)

        service.requestServerSelection()
        XCTAssertEqual(service.activeMatch?.needsServerSelection, true)
        XCTAssertEqual(store.active?.needsServerSelection, true)

        service.selectServer(.right)
        XCTAssertEqual(service.activeMatch?.currentServer, .right)
        service.awardPoint(to: .left)
        XCTAssertEqual(service.activeMatch?.currentServer, .right)
        XCTAssertEqual(store.active?.currentGame.leftPoints, 1)
    }

    func testStartMatchPersistsDeuceFormat() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        var settings = MatchSettings.default
        settings.deuceFormat = .silverPoint
        service.startMatch(settings: settings)
        XCTAssertEqual(service.activeMatch?.settings.deuceFormat, .silverPoint)
        XCTAssertEqual(store.active?.settings.deuceFormat, .silverPoint)
    }

    func testDeuceFormatCanBeChangedMidMatchAndPersists() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        var settings = MatchSettings.default
        settings.deuceFormat = .advantage
        service.startMatch(settings: settings)
        service.selectServer(.left)
        for _ in 0..<3 {
            service.awardPoint(to: .left)
            service.awardPoint(to: .right)
        }
        XCTAssertEqual(service.activeMatch?.gameStatusLine, "Deuce")

        service.setDeuceFormat(.goldenPoint)
        XCTAssertEqual(service.activeMatch?.settings.deuceFormat, .goldenPoint)
        XCTAssertEqual(service.activeMatch?.currentGame.isGoldenPointActive, true)
        XCTAssertEqual(store.active?.settings.deuceFormat, .goldenPoint)

        service.awardPoint(to: .left)
        XCTAssertEqual(service.activeMatch?.currentSet.leftGames, 1)
    }

    func testDeuceFormatChangeIgnoredWithoutAnActiveMatch() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.setDeuceFormat(.advantage)
        XCTAssertNil(service.activeMatch)
    }

    func testExpireInactiveMatchWithPointsEndsEarly() throws {
        let engine = ScoringEngine()
        let oldDate = Date(timeIntervalSinceNow: -31 * 60)
        var match = engine.startMatch(at: oldDate)
        match = try engine.apply(.selectServer(.left), to: match, at: oldDate)
        match = try engine.apply(.pointWon(.left), to: match, at: oldDate)

        let service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        XCTAssertEqual(service.activeMatch?.status, .endedEarly)
        XCTAssertEqual(service.archivedMatches.count, 1)
    }

    func testExpireInactiveMatchWithRecentPointStaysInProgress() throws {
        let engine = ScoringEngine()
        let startDate = Date(timeIntervalSinceNow: -60 * 60)
        let recentDate = Date(timeIntervalSinceNow: -5 * 60)
        var match = engine.startMatch(at: startDate)
        match = try engine.apply(.selectServer(.left), to: match, at: startDate)
        match = try engine.apply(.pointWon(.left), to: match, at: recentDate)

        let service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        XCTAssertEqual(service.activeMatch?.status, .inProgress)
        XCTAssertTrue(service.archivedMatches.isEmpty)
    }

    func testExpireInactiveZeroPointMatchDiscards() throws {
        let engine = ScoringEngine()
        let oldDate = Date(timeIntervalSinceNow: -31 * 60)
        let match = engine.startMatch(at: oldDate)

        let service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        XCTAssertNil(service.activeMatch)
        XCTAssertTrue(service.archivedMatches.isEmpty)
    }

    func testExpireInactiveRecentZeroPointMatchStaysInProgress() throws {
        let engine = ScoringEngine()
        let recentDate = Date(timeIntervalSinceNow: -5 * 60)
        let match = engine.startMatch(at: recentDate)

        let service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        XCTAssertEqual(service.activeMatch?.status, .inProgress)
    }

    func testRestoreExpiresInactiveMatch() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        let oldDate = Date(timeIntervalSinceNow: -31 * 60)
        var match = engine.startMatch(at: oldDate)
        match = try engine.apply(.selectServer(.left), to: match, at: oldDate)
        match = try engine.apply(.pointWon(.left), to: match, at: oldDate)
        store.active = match

        let service = MatchService(store: store)

        XCTAssertEqual(service.activeMatch?.status, .endedEarly)
        XCTAssertEqual(service.archivedMatches.count, 1)
    }

    func testUndoAllPointsOnInactiveMatchDiscards() throws {
        let engine = ScoringEngine()
        let startDate = Date(timeIntervalSinceNow: -31 * 60)
        let recentDate = Date(timeIntervalSinceNow: -5 * 60)
        var match = engine.startMatch(at: startDate)
        match = try engine.apply(.selectServer(.left), to: match, at: startDate)
        match = try engine.apply(.pointWon(.left), to: match, at: recentDate)

        let service = serviceWithActiveMatch(match)
        service.undoLastPoint()

        XCTAssertNil(service.activeMatch)
        XCTAssertTrue(service.archivedMatches.isEmpty)
    }

    func testDeleteArchivedMatchRemovesItAndRecordsTombstone() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        service.endMatchEarly()
        service.acknowledgeCompletedMatch()
        let id = try! XCTUnwrap(service.archivedMatches.first?.id)

        service.deleteArchivedMatch(id: id)

        XCTAssertTrue(service.archivedMatches.isEmpty)
        XCTAssertTrue(store.archive.isEmpty)
        XCTAssertEqual(store.deletedIDs, [id])
    }

    func testDeletedMatchDoesNotReturnViaRemoteSnapshot() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        var match = engine.startMatch()
        match = try engine.apply(.selectServer(.left), to: match)
        match = try engine.apply(.pointWon(.left), to: match)
        match = try engine.apply(.endEarly, to: match)
        store.archive = [match]

        let service = MatchService(store: store)
        service.deleteArchivedMatch(id: match.id)

        // The Watch is unaware and pushes its full archive back.
        service.applyRemoteSnapshot(active: nil, archive: [match])

        XCTAssertTrue(service.archivedMatches.isEmpty)
        XCTAssertTrue(store.archive.isEmpty)
    }

    func testDeletedMatchStaysDeletedAcrossRestore() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        var match = engine.startMatch()
        match = try engine.apply(.endEarly, to: match)
        store.archive = [match]

        let service = MatchService(store: store)
        service.deleteArchivedMatch(id: match.id)

        let restored = MatchService(store: store)
        XCTAssertTrue(restored.archivedMatches.isEmpty)
        XCTAssertEqual(restored.deletedMatchIDs, [match.id])
    }

    func testDeleteIgnoresTheActiveMatch() {
        let store = InMemoryMatchStore()
        let service = MatchService(store: store)
        service.startMatch()
        service.selectServer(.left)
        service.awardPoint(to: .left)
        service.endMatchEarly()
        let id = try! XCTUnwrap(service.activeMatch?.id)

        service.deleteArchivedMatch(id: id)

        XCTAssertEqual(service.archivedMatches.count, 1)
        XCTAssertTrue(service.deletedMatchIDs.isEmpty)
    }

    func testRemoteDeletionsPruneArchiveButNotActiveMatch() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        var archived = engine.startMatch()
        archived = try engine.apply(.endEarly, to: archived)
        store.archive = [archived]

        let service = MatchService(store: store)
        service.startMatch()
        let activeID = try XCTUnwrap(service.activeMatch?.id)

        service.applyRemoteDeletions([archived.id, activeID])

        XCTAssertTrue(service.archivedMatches.isEmpty)
        XCTAssertTrue(store.archive.isEmpty)
        XCTAssertEqual(service.activeMatch?.id, activeID)
        XCTAssertEqual(service.deletedMatchIDs, [archived.id])
        XCTAssertEqual(store.deletedIDs, [archived.id])
    }

    func testDeletedIDsSurviveFileStoreRoundTrip() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = FileMatchStore(directory: dir)
        XCTAssertTrue(try store.loadDeletedMatchIDs().isEmpty)

        let ids: Set<UUID> = [UUID(), UUID()]
        try store.saveDeletedMatchIDs(ids)

        XCTAssertEqual(try FileMatchStore(directory: dir).loadDeletedMatchIDs(), ids)
    }

    func testNoteIsStoredTrimmedAndClearedWhenBlank() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        let match = try engine.apply(.endEarly, to: engine.startMatch())
        store.archive = [match]
        let service = MatchService(store: store)

        service.setNote("  Played with Sam, windy court\n", for: match.id)
        XCTAssertEqual(service.note(for: match.id), "Played with Sam, windy court")
        XCTAssertEqual(store.notes[match.id], "Played with Sam, windy court")

        service.setNote("   ", for: match.id)
        XCTAssertEqual(service.note(for: match.id), "")
        XCTAssertNil(store.notes[match.id])
    }

    func testNoteSurvivesRemoteSnapshotAndRestore() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        let match = try engine.apply(.endEarly, to: engine.startMatch())
        store.archive = [match]
        let service = MatchService(store: store)
        service.setNote("Left knee sore", for: match.id)

        service.applyRemoteSnapshot(active: nil, archive: [match])
        XCTAssertEqual(service.note(for: match.id), "Left knee sore")

        let restored = MatchService(store: store)
        XCTAssertEqual(restored.note(for: match.id), "Left knee sore")
    }

    func testDeletingAMatchDropsItsNoteForGood() throws {
        let store = InMemoryMatchStore()
        let engine = ScoringEngine()
        let match = try engine.apply(.endEarly, to: engine.startMatch())
        store.archive = [match]
        let service = MatchService(store: store)
        service.setNote("Best of the season", for: match.id)

        service.deleteArchivedMatch(id: match.id)
        XCTAssertTrue(store.notes.isEmpty)

        // A detail view saving its draft on dismissal must not bring the note back.
        service.setNote("Best of the season", for: match.id)
        XCTAssertTrue(service.matchNotes.isEmpty)
    }

    func testNotesSurviveFileStoreRoundTrip() throws {
        let dir = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: dir) }
        let store = FileMatchStore(directory: dir)
        XCTAssertTrue(try store.loadMatchNotes().isEmpty)

        let notes = [UUID(): "First note", UUID(): "Second note"]
        try store.saveMatchNotes(notes)

        XCTAssertEqual(try FileMatchStore(directory: dir).loadMatchNotes(), notes)
    }

    private func serviceWithActiveMatch(_ match: MatchState) -> MatchService {
        let store = InMemoryMatchStore()
        store.active = match
        return MatchService(store: store)
    }
}

File: android/domain/src/test/kotlin/com/wristrally/domain/MatchServiceTests.kt
package com.wristrally.domain

import java.nio.file.Files
import java.time.Instant
import java.util.UUID
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path

class MatchServiceTests {
    @Test
    fun testPersistAndRestoreActiveMatch() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        assertEquals(1, store.active?.currentGame?.leftPoints)

        val restored = MatchService(store)
        assertEquals(1, restored.activeMatch?.currentGame?.leftPoints)
        assertEquals(service.activeMatch?.id, restored.activeMatch?.id)
    }

    @Test
    fun testFileStoreRoundTrip(@TempDir dir: Path) {
        val store = FileMatchStore(dir)
        val engine = ScoringEngine()
        var match = engine.startMatch()
        match = engine.apply(MatchAction.SelectServer(Side.Left), match)
        match = engine.apply(MatchAction.PointWon(Side.Right), match)
        store.saveActiveMatch(match)

        val loaded = store.loadActiveMatch()
        assertEquals(1, loaded?.currentGame?.rightPoints)

        match = engine.apply(MatchAction.EndEarly, match)
        store.saveActiveMatch(null)
        store.archiveMatch(match)
        val archive = store.loadArchivedMatches()
        assertEquals(1, archive.size)
        assertEquals(MatchStatus.EndedEarly, archive.first().status)
    }

    @Test
    fun testDiscardDoesNotArchive() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        service.discardMatch()
        assertNull(service.activeMatch)
        assertTrue(service.archivedMatches.isEmpty())
        assertNull(store.active)
    }

    @Test
    fun testDiscardBeforeServerSelectionClearsActiveMatch() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        assertEquals(true, service.activeMatch?.isWaitingForFirstServe)
        service.discardMatch()
        assertNull(service.activeMatch)
        assertTrue(service.archivedMatches.isEmpty())
        assertNull(store.active)
    }

    @Test
    fun testEndEarlyArchivesAndKeepsActiveUntilAck() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        service.endMatchEarly()
        assertEquals(MatchStatus.EndedEarly, service.activeMatch?.status)
        assertEquals(1, service.archivedMatches.size)
        service.acknowledgeCompletedMatch()
        assertNull(service.activeMatch)
        assertEquals(1, service.archivedMatches.size)
    }

    @Test
    fun testUndoViaService() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        service.awardPoint(Side.Right)
        service.undoLastPoint()
        assertEquals(1, service.activeMatch?.currentGame?.leftPoints)
        assertEquals(0, service.activeMatch?.currentGame?.rightPoints)
    }

    @Test
    fun testSelectingServerUpdatesActiveMatch() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        assertEquals(true, service.activeMatch?.needsServerSelection)
        service.selectServer(Side.Right)
        assertEquals(false, service.activeMatch?.needsServerSelection)
        assertEquals(Side.Right, service.activeMatch?.currentServer)
    }

    @Test
    fun testCompleteWarmUpPersistsAndKeepsServePrompt() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        assertEquals(true, service.activeMatch?.needsWarmUp)
        assertEquals(true, service.activeMatch?.needsServerSelection)
        service.completeWarmUp()
        assertEquals(false, service.activeMatch?.needsWarmUp)
        assertEquals(true, service.activeMatch?.needsServerSelection)
        assertEquals(false, store.active?.needsWarmUp)
    }

    @Test
    fun testNewServeAtTheChangeoverPromptsAndPersists() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch(MatchSettings().copy(setsToWin = 2))
        service.selectServer(Side.Left)
        repeat(24) { service.awardPoint(Side.Left) }
        assertEquals(1, service.activeMatch?.completedSets?.size)

        service.requestServerSelection()
        assertEquals(true, service.activeMatch?.needsServerSelection)
        assertEquals(true, store.active?.needsServerSelection)

        service.selectServer(Side.Right)
        assertEquals(Side.Right, service.activeMatch?.currentServer)
        service.awardPoint(Side.Left)
        assertEquals(Side.Right, service.activeMatch?.currentServer)
        assertEquals(1, store.active?.currentGame?.leftPoints)
    }

    @Test
    fun testStartMatchPersistsDeuceFormat() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch(MatchSettings().copy(deuceFormat = DeuceFormat.SilverPoint))
        assertEquals(DeuceFormat.SilverPoint, service.activeMatch?.settings?.deuceFormat)
        assertEquals(DeuceFormat.SilverPoint, store.active?.settings?.deuceFormat)
    }

    @Test
    fun testDeuceFormatCanBeChangedMidMatchAndPersists() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch(MatchSettings().copy(deuceFormat = DeuceFormat.Advantage))
        service.selectServer(Side.Left)
        repeat(3) {
            service.awardPoint(Side.Left)
            service.awardPoint(Side.Right)
        }
        assertEquals("Deuce", service.activeMatch?.gameStatusLine)

        service.setDeuceFormat(DeuceFormat.GoldenPoint)
        assertEquals(DeuceFormat.GoldenPoint, service.activeMatch?.settings?.deuceFormat)
        assertEquals(true, service.activeMatch?.currentGame?.isGoldenPointActive)
        assertEquals(DeuceFormat.GoldenPoint, store.active?.settings?.deuceFormat)

        service.awardPoint(Side.Left)
        assertEquals(1, service.activeMatch?.currentSet?.leftGames)
    }

    @Test
    fun testDeuceFormatChangeIgnoredWithoutAnActiveMatch() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.setDeuceFormat(DeuceFormat.Advantage)
        assertNull(service.activeMatch)
    }

    @Test
    fun testExpireInactiveMatchWithPointsEndsEarly() {
        val engine = ScoringEngine()
        val oldDate = Instant.now().minusSeconds(31 * 60)
        var match = engine.startMatch(at = oldDate)
        match = engine.apply(MatchAction.SelectServer(Side.Left), match, oldDate)
        match = engine.apply(MatchAction.PointWon(Side.Left), match, oldDate)

        val service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        assertEquals(MatchStatus.EndedEarly, service.activeMatch?.status)
        assertEquals(1, service.archivedMatches.size)
    }

    @Test
    fun testExpireInactiveMatchWithRecentPointStaysInProgress() {
        val engine = ScoringEngine()
        val startDate = Instant.now().minusSeconds(60 * 60)
        val recentDate = Instant.now().minusSeconds(5 * 60)
        var match = engine.startMatch(at = startDate)
        match = engine.apply(MatchAction.SelectServer(Side.Left), match, startDate)
        match = engine.apply(MatchAction.PointWon(Side.Left), match, recentDate)

        val service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        assertEquals(MatchStatus.InProgress, service.activeMatch?.status)
        assertTrue(service.archivedMatches.isEmpty())
    }

    @Test
    fun testExpireInactiveZeroPointMatchDiscards() {
        val engine = ScoringEngine()
        val oldDate = Instant.now().minusSeconds(31 * 60)
        val match = engine.startMatch(at = oldDate)

        val service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        assertNull(service.activeMatch)
        assertTrue(service.archivedMatches.isEmpty())
    }

    @Test
    fun testExpireInactiveRecentZeroPointMatchStaysInProgress() {
        val engine = ScoringEngine()
        val recentDate = Instant.now().minusSeconds(5 * 60)
        val match = engine.startMatch(at = recentDate)

        val service = serviceWithActiveMatch(match)
        service.expireInactiveMatchIfNeeded()

        assertEquals(MatchStatus.InProgress, service.activeMatch?.status)
    }

    @Test
    fun testRestoreExpiresInactiveMatch() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        val oldDate = Instant.now().minusSeconds(31 * 60)
        var match = engine.startMatch(at = oldDate)
        match = engine.apply(MatchAction.SelectServer(Side.Left), match, oldDate)
        match = engine.apply(MatchAction.PointWon(Side.Left), match, oldDate)
        store.active = match

        val service = MatchService(store)

        assertEquals(MatchStatus.EndedEarly, service.activeMatch?.status)
        assertEquals(1, service.archivedMatches.size)
    }

    @Test
    fun testUndoAllPointsOnInactiveMatchDiscards() {
        val engine = ScoringEngine()
        val startDate = Instant.now().minusSeconds(31 * 60)
        val recentDate = Instant.now().minusSeconds(5 * 60)
        var match = engine.startMatch(at = startDate)
        match = engine.apply(MatchAction.SelectServer(Side.Left), match, startDate)
        match = engine.apply(MatchAction.PointWon(Side.Left), match, recentDate)

        val service = serviceWithActiveMatch(match)
        service.undoLastPoint()

        assertNull(service.activeMatch)
        assertTrue(service.archivedMatches.isEmpty())
    }

    @Test
    fun testDeleteArchivedMatchRemovesItAndRecordsTombstone() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        service.endMatchEarly()
        service.acknowledgeCompletedMatch()
        val id = service.archivedMatches.first().id

        service.deleteArchivedMatch(id)

        assertTrue(service.archivedMatches.isEmpty())
        assertTrue(store.archive.isEmpty())
        assertEquals(setOf(id), store.deletedIDs)
    }

    @Test
    fun testDeletedMatchDoesNotReturnViaRemoteSnapshot() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        var match = engine.startMatch()
        match = engine.apply(MatchAction.SelectServer(Side.Left), match)
        match = engine.apply(MatchAction.PointWon(Side.Left), match)
        match = engine.apply(MatchAction.EndEarly, match)
        store.archive = mutableListOf(match)

        val service = MatchService(store)
        service.deleteArchivedMatch(match.id)

        service.applyRemoteSnapshot(null, listOf(match))

        assertTrue(service.archivedMatches.isEmpty())
        assertTrue(store.archive.isEmpty())
    }

    @Test
    fun testDeletedMatchStaysDeletedAcrossRestore() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        var match = engine.startMatch()
        match = engine.apply(MatchAction.EndEarly, match)
        store.archive = mutableListOf(match)

        val service = MatchService(store)
        service.deleteArchivedMatch(match.id)

        val restored = MatchService(store)
        assertTrue(restored.archivedMatches.isEmpty())
        assertEquals(setOf(match.id), restored.deletedMatchIDs)
    }

    @Test
    fun testDeleteIgnoresTheActiveMatch() {
        val store = InMemoryMatchStore()
        val service = MatchService(store)
        service.startMatch()
        service.selectServer(Side.Left)
        service.awardPoint(Side.Left)
        service.endMatchEarly()
        val id = service.activeMatch!!.id

        service.deleteArchivedMatch(id)

        assertEquals(1, service.archivedMatches.size)
        assertTrue(service.deletedMatchIDs.isEmpty())
    }

    @Test
    fun testRemoteDeletionsPruneArchiveButNotActiveMatch() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        var archived = engine.startMatch()
        archived = engine.apply(MatchAction.EndEarly, archived)
        store.archive = mutableListOf(archived)

        val service = MatchService(store)
        service.startMatch()
        val activeID = service.activeMatch!!.id

        service.applyRemoteDeletions(setOf(archived.id, activeID))

        assertTrue(service.archivedMatches.isEmpty())
        assertTrue(store.archive.isEmpty())
        assertEquals(activeID, service.activeMatch?.id)
        assertEquals(setOf(archived.id), service.deletedMatchIDs)
        assertEquals(setOf(archived.id), store.deletedIDs)
    }

    @Test
    fun testDeletedIDsSurviveFileStoreRoundTrip(@TempDir dir: Path) {
        val store = FileMatchStore(dir)
        assertTrue(store.loadDeletedMatchIDs().isEmpty())

        val ids = setOf(UUID.randomUUID(), UUID.randomUUID())
        store.saveDeletedMatchIDs(ids)

        assertEquals(ids, FileMatchStore(dir).loadDeletedMatchIDs())
    }

    @Test
    fun testNoteIsStoredTrimmedAndClearedWhenBlank() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        val match = engine.apply(MatchAction.EndEarly, engine.startMatch())
        store.archive = mutableListOf(match)
        val service = MatchService(store)

        service.setNote("  Played with Sam, windy court\n", match.id)
        assertEquals("Played with Sam, windy court", service.note(match.id))
        assertEquals("Played with Sam, windy court", store.notes[match.id])

        service.setNote("   ", match.id)
        assertEquals("", service.note(match.id))
        assertNull(store.notes[match.id])
    }

    @Test
    fun testNoteSurvivesRemoteSnapshotAndRestore() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        val match = engine.apply(MatchAction.EndEarly, engine.startMatch())
        store.archive = mutableListOf(match)
        val service = MatchService(store)
        service.setNote("Left knee sore", match.id)

        service.applyRemoteSnapshot(null, listOf(match))
        assertEquals("Left knee sore", service.note(match.id))

        val restored = MatchService(store)
        assertEquals("Left knee sore", restored.note(match.id))
    }

    @Test
    fun testDeletingAMatchDropsItsNoteForGood() {
        val store = InMemoryMatchStore()
        val engine = ScoringEngine()
        val match = engine.apply(MatchAction.EndEarly, engine.startMatch())
        store.archive = mutableListOf(match)
        val service = MatchService(store)
        service.setNote("Best of the season", match.id)

        service.deleteArchivedMatch(match.id)
        assertTrue(store.notes.isEmpty())

        service.setNote("Best of the season", match.id)
        assertTrue(service.matchNotes.isEmpty())
    }

    @Test
    fun testNotesSurviveFileStoreRoundTrip(@TempDir dir: Path) {
        val store = FileMatchStore(dir)
        assertTrue(store.loadMatchNotes().isEmpty())

        val notes = mapOf(UUID.randomUUID() to "First note", UUID.randomUUID() to "Second note")
        store.saveMatchNotes(notes)

        assertEquals(notes, FileMatchStore(dir).loadMatchNotes())
    }

    private fun serviceWithActiveMatch(match: MatchState): MatchService {
        val store = InMemoryMatchStore()
        store.active = match
        return MatchService(store)
    }
}
